### DISCUSSION TITLE
[Design] Core Abstractions

### DISCUSSION THREAD

**Discussion by andyk** [agentos-project/agentos#112] [Designs]
This discussion thread is about the design of AgentOS core abstractions.

Design effort leader: @nickjalbert
Design Doc Pull Request: #3

Versions of the design doc:
* [v1](https://github.com/agentos-project/design_docs/blob/d8a309a5a89237edcda24cca838859398391c269/abstractions.rst)
* [v2](https://github.com/agentos-project/design_docs/blob/47c4251741f37cbd522c31e30031cbe4d8580361/abstractions.rst)
* [v3](https://github.com/agentos-project/design_docs/blob/89f507c373b296ec252971d585d919657543eaab/abstractions.rst)

**Comment by andyk** [comment]
The following design review is on [v2 of the design doc](https://github.com/agentos-project/design_docs/blob/47c4251741f37cbd522c31e30031cbe4d8580361/abstractions.rst).

* In the Agent class, do we want `train()` and `advance()` to contain `pass` or `raise NotImplementedError`?

* Should our core-most ("base") abstrations be Python [Abstract Base Classes (ABCs)](https://docs.python.org/3/library/abc.html) (see also [the ABC PEP](https://www.python.org/dev/peps/pep-3119/))? I don't have a lot of experience with ABCs, so I'm not sure if they are a fit.

* ACS -> ACR

* Is the current design future proof enough for us to be able to separate out Reward Function as another core abstraction in the near-ish future? Should we add a note about it somewhere (under the section on Environment perhaps)?

* I can imaging AgentOS users getting confused by `Agent` and `Trainer` both having a `.train()` function. Here are a few (non mutally exclusive) ideas to address this:
  * Idea 1: drop Trainer as a separate concept/class (for now), and instead allow Policies to optionally support a `.train()` or `.improve()` method. This would also simplify the project by reducing the number of core abstractions by 25%.
  * Idea 2: different function names for Agent and Trainer (or Policy if we implement Idea 1) => `Agent.learn()` and `Trainer.train()` (or `Policy.train()` or `Policy.improve()`).
    * This would align the architecture/API better with our convention of referring to agents that can learn as "learning agents".
    * Also, "training" tends to invoke the idea of supervised learning, whereas "learning" invokes "reinforcement learning".
  * Idea 3: Have AgentOS train an agent by working directly with its policy object (e.g., by calling `agent_instance.policy.improve()` instead of doing it via a pass-through member function of agent.  

* "This file will be managed by the AgentOS Component System (ACS) to allow for easy composition and reuse of AgentOS components." Maybe we can add a note after this along the lines of: "Developers can also directly edit agent.ini if they prefer."

* Do we envision adding other sub-classes of these base classes to the `agentos` package (e.g., [agentos.agents.RandomAgent](https://github.com/agentos-project/agentos/blob/master/agentos/agents.py)) or do we expect all non base agents to come from ACR? (I think I'm in favor of the latter).

**Reply by nickjalbert** [reply]
> In the Agent class, do we want train() and advance() to contain pass or raise NotImplementedError?

Split the difference?  It seems like it's reasonable for an agent to have a no-op `train()`, but maybe we require you to implement something for `advance()`?  I think I could be convinced either way, though.

> Should our core-most ("base") abstractions be Python Abstract Base Classes (ABCs) (see also the ABC PEP)?

I also don't have much experience with ABCs.  My understanding from a quick glance at the docs is that it allows you to force subclasses to define methods you've marked as abstract.  This could be useful, but perhaps we punt until everything is a little more stable.  I'll add a note under the long term plans section.

>   separate out Reward Function as another core abstraction

This seems like it might be useful.  I'll add a note under long term plans.

> AgentOS users getting confused by Agent and Trainer both having a .train() function

I agree this is inelegant.  My initial theory for keeping Trainer separate is that you may want to try to train the same policy using different training methods (but I'm having a bit of difficulty coming up with a good example off the top of my head).  Maybe we collapse this for the moment and see if we bump into a case where it makes sense to have a separate Trainer.

Somewhat related: the Agent API is a bit inconsistent in that `advance` is a single step whereas `train` is potentially-many steps.  I don't have a great solution for this (and I'm not even sure this is really a problem), but it's something to think on.

>  add a note after this along the lines of: "Developers can also directly edit agent.ini if they prefer."

done

> Do we envision adding other sub-classes of these base classes to the agentos package (e.g., agentos.agents.RandomAgent) or do we expect all non base agents to come from ACR? (I think I'm in favor of the latter).

I think `agentos init` should generate a minimal runnable agent, but we should eventually separate everything else out from the core of AgentOS (and use ACR).

## Major changes

* `train()` is by default a no-op, `advance()` by default raises NotImplementedError
* Drop `Trainer`, Policy gets a `improve()` method and Agent gets a `learn()` method (instead of `train()`).

**Reply by nickjalbert** [reply]
I pushed an update to the [PR](https://github.com/agentos-project/design_docs/pull/3) making the above major changes.  Happy to revert some or all as the discussion proceeds.

**Reply by nickjalbert** [reply]
Current version: [v3](https://github.com/agentos-project/design_docs/blob/89f507c373b296ec252971d585d919657543eaab/abstractions.rst)

**Reply by andyk** [reply]
> 
> 
> > In the Agent class, do we want train() and advance() to contain pass or raise NotImplementedError?
> 
> Split the difference? It seems like it's reasonable for an agent to have a no-op `train()`, but maybe we require you to implement something for `advance()`? I think I could be convinced either way, though.
> 

Yeah I agree: `pass` for `train()`, `raise NotImplementedError` for `advance()`.

> > Should our core-most ("base") abstractions be Python Abstract Base Classes (ABCs) (see also the ABC PEP)?
> 
> I also don't have much experience with ABCs. My understanding from a quick glance at the docs is that it allows you to force subclasses to define methods you've marked as abstract. This could be useful, but perhaps we punt until everything is a little more stable. I'll add a note under the long term plans section.

Sounds good.
 
> > separate out Reward Function as another core abstraction
> 
> This seems like it might be useful. I'll add a note under long term plans.
> 
> > AgentOS users getting confused by Agent and Trainer both having a .train() function
> 
> I agree this is inelegant. My initial theory for keeping Trainer separate is that you may want to try to train the same policy using different training methods (but I'm having a bit of difficulty coming up with a good example off the top of my head). Maybe we collapse this for the moment and see if we bump into a case where it makes sense to have a separate Trainer.

Sounds good.

> Somewhat related: the Agent API is a bit inconsistent in that `advance` is a single step whereas `train` is potentially-many steps. I don't have a great solution for this (and I'm not even sure this is really a problem), but it's something to think on.

In my head I've been separating the notion of an agent taking a step (i.e., in their world) from the notion of an agent advancing (i.e, progressing through time, probably changing some how as they do).

Separating these concepts makes it possible to have `advance()` call `self.env.step()` less frequently than every time it runs (e.g., every 10th time advance runs, it could call step) or more frequently (e.g., advance calls step twice). 

Hopefully this provides enough flexibility for the agent developer with a rather minimal API.

The more familiar template wherein every time an agent advances it takes one env.step but does many iterations of learning (i.e., many rollouts) is similar to planning (as in [David Silver lecture 3](https://www.davidsilver.uk/wp-content/uploads/2020/03/DP.pdf)) and model-based RL, when an agent spends time thinking (unconsciously or consciously) before each action.

In general, I think the agent dev can independently decide the m-to-n relationship between `advance()` and each of `env.step()`, `policy.decide()`, `policy.train()`.

I haven't thought about it a ton, but the dev can also adjust how many iterations `env.step()`, `policy.decide()`, `policy.train()` each perform of their core functionality. For example, they use or write a policy that does one step of training per call to `.train()`.

> > add a note after this along the lines of: "Developers can also directly edit agent.ini if they prefer."
> 
> done
> 
> > Do we envision adding other sub-classes of these base classes to the agentos package (e.g., agentos.agents.RandomAgent) or do we expect all non base agents to come from ACR? (I think I'm in favor of the latter).
> 
> I think `agentos init` should generate a minimal runnable agent, but we should eventually separate everything else out from the core of AgentOS (and use ACR).

SGTM

> ## Major changes
> 
> * `train()` is by default a no-op, `advance()` by default raises NotImplementedError
> 
> * Drop `Trainer`, Policy gets a `improve()` method and Agent gets a `learn()` method (instead of `train()`).

**Comment by andyk** [comment]
After doing our [survey of the existing Deep RL frameworks](https://docs.google.com/presentation/d/1xe-6FjY6ryHGHVG8MJimm3YmGze2WY2RMSBmXKS2HBo/edit) (now 14, and counting), we decided there is a place for a stand-alone Trainer in the core abstractions after-all.

We also may want to add a core abstraction for MetaTrainer to support the increasingly popular techniques from Meta-RL
